File: pricing-service-java/src/main/java/com/example/retail/javaapp/service/PriceQuoteService.java
package com.example.retail.javaapp.service;

import com.example.retail.contract.PriceQuoteRequest;
import com.example.retail.contract.PriceQuoteResponse;
import com.example.retail.contract.PriceSource;
import com.example.retail.contract.ProductPrice;
import com.example.retail.javaapp.cache.CacheKey;
import com.example.retail.javaapp.cache.QuoteComputation;
import com.example.retail.javaapp.cache.RedisBackedQuoteCache;
import com.example.retail.javaapp.error.ProductNotFoundException;
import com.example.retail.javaapp.repository.ProductCatalogRepository;
import com.example.retail.javaapp.repository.PromotionRepository;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;
import org.springframework.stereotype.Service;

@Service
public class PriceQuoteService {

    private static final Logger log = LoggerFactory.getLogger(PriceQuoteService.class);

    private final ProductCatalogRepository productCatalogRepository;
    private final PromotionRepository promotionRepository;
    private final PriceQuoteCalculator priceQuoteCalculator;
    private final RedisBackedQuoteCache quoteCache;
    private final MeterRegistry meterRegistry;

    public PriceQuoteService(
            ProductCatalogRepository productCatalogRepository,
            PromotionRepository promotionRepository,
            PriceQuoteCalculator priceQuoteCalculator,
            RedisBackedQuoteCache quoteCache,
            MeterRegistry meterRegistry
    ) {
        this.productCatalogRepository = productCatalogRepository;
        this.promotionRepository = promotionRepository;
        this.priceQuoteCalculator = priceQuoteCalculator;
        this.quoteCache = quoteCache;
        this.meterRegistry = meterRegistry;
    }

    public PriceQuoteResponse quote(PriceQuoteRequest request) {
        Timer.Sample sample = Timer.start(meterRegistry);
        String cacheKey = CacheKey.from(request);
        String traceId = currentTraceId();

        return quoteCache.find(cacheKey)
                .map(computation -> {
                    long latencyMs = stopTimer(sample, "CACHE");
                    log.info("quote_cache_hit sku={} traceId={} latencyMs={}", request.sku(), traceId, latencyMs);
                    return toResponse(request, computation, PriceSource.CACHE, traceId, latencyMs);
                })
                .orElseGet(() -> computeFreshQuote(request, cacheKey, traceId, sample));
    }

    private PriceQuoteResponse computeFreshQuote(
            PriceQuoteRequest request,
            String cacheKey,
            String traceId,
            Timer.Sample sample
    ) {
        ProductPrice product = productCatalogRepository.findBySkuAndStoreId(request.sku(), request.storeId())
                .orElseThrow(() -> new ProductNotFoundException(request.sku(), request.storeId()));

        QuoteComputation computation = priceQuoteCalculator.compute(
                product,
                promotionRepository.findBySku(request.sku()),
                request
        );
        quoteCache.store(cacheKey, computation);

        long latencyMs = stopTimer(sample, "DATABASE");
        log.info("quote_computed sku={} traceId={} latencyMs={}", request.sku(), traceId, latencyMs);
        return toResponse(request, computation, PriceSource.DATABASE, traceId, latencyMs);
    }

    private PriceQuoteResponse toResponse(
            PriceQuoteRequest request,
            QuoteComputation computation,
            PriceSource source,
            String traceId,
            long latencyMs
    ) {
        return new PriceQuoteResponse(
                request.sku(),
                request.storeId(),
                computation.basePrice(),
                computation.appliedPromotions(),
                computation.finalPrice(),
                computation.fulfillmentOptions(),
                source,
                traceId,
                latencyMs
        );
    }

    private long stopTimer(Timer.Sample sample, String source) {
        return Math.round(sample.stop(Timer.builder("pricing.quote.latency")
                .tag("source", source)
                .register(meterRegistry)) / 1_000_000.0d);
    }

    private String currentTraceId() {
        String traceId = MDC.get("traceId");
        return traceId == null ? "missing-trace-id" : traceId;
    }
}

File: pricing-service-kotlin/src/main/kotlin/com/example/retail/kotlinapp/service/PriceQuoteService.kt
package com.example.retail.kotlinapp.service

import com.example.retail.contract.PriceQuoteRequest
import com.example.retail.contract.PriceQuoteResponse
import com.example.retail.contract.PriceSource
import com.example.retail.kotlinapp.cache.CacheKey
import com.example.retail.kotlinapp.cache.QuoteComputation
import com.example.retail.kotlinapp.cache.RedisBackedQuoteCache
import com.example.retail.kotlinapp.error.ProductNotFoundException
import com.example.retail.kotlinapp.repository.ProductCatalogRepository
import com.example.retail.kotlinapp.repository.PromotionRepository
import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Timer
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import org.springframework.stereotype.Service
import kotlin.math.roundToLong

@Service
class PriceQuoteService(
    private val productCatalogRepository: ProductCatalogRepository,
    private val promotionRepository: PromotionRepository,
    private val priceQuoteCalculator: PriceQuoteCalculator,
    private val quoteCache: RedisBackedQuoteCache,
    private val meterRegistry: MeterRegistry,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    fun quote(request: PriceQuoteRequest): PriceQuoteResponse {
        val sample = Timer.start(meterRegistry)
        val cacheKey = CacheKey.from(request)
        val traceId = MDC.get("traceId") ?: "missing-trace-id"

        return quoteCache.find(cacheKey)?.let { computation ->
            val latencyMs = stopTimer(sample, "CACHE")
            log.info("quote_cache_hit sku={} traceId={} latencyMs={}", request.sku(), traceId, latencyMs)
            toResponse(request, computation, PriceSource.CACHE, traceId, latencyMs)
        } ?: computeFreshQuote(request, cacheKey, traceId, sample)
    }

    private fun computeFreshQuote(
        request: PriceQuoteRequest,
        cacheKey: String,
        traceId: String,
        sample: Timer.Sample,
    ): PriceQuoteResponse {
        val product = productCatalogRepository.findBySkuAndStoreId(request.sku(), request.storeId())
            ?: throw ProductNotFoundException(request.sku(), request.storeId())

        val computation = priceQuoteCalculator.compute(product, promotionRepository.findBySku(request.sku()), request)
        quoteCache.store(cacheKey, computation)

        val latencyMs = stopTimer(sample, "DATABASE")
        log.info("quote_computed sku={} traceId={} latencyMs={}", request.sku(), traceId, latencyMs)
        return toResponse(request, computation, PriceSource.DATABASE, traceId, latencyMs)
    }

    private fun toResponse(
        request: PriceQuoteRequest,
        computation: QuoteComputation,
        source: PriceSource,
        traceId: String,
        latencyMs: Long,
    ): PriceQuoteResponse = PriceQuoteResponse(
        request.sku(),
        request.storeId(),
        computation.basePrice,
        computation.appliedPromotions,
        computation.finalPrice,
        computation.fulfillmentOptions,
        source,
        traceId,
        latencyMs,
    )

    private fun stopTimer(sample: Timer.Sample, source: String): Long =
        (sample.stop(Timer.builder("pricing.quote.latency").tag("source", source).register(meterRegistry)) / 1_000_000.0).roundToLong()
}
